### ISSUE TITLE
Graphite index causing nginx timeouts

### ISSUE THREAD

**Issue by ivaniliev359** [graphite-project/graphite-web#2279]
Hello,

Quite often when I'm browsing the metrics tree from the graphite webapp, the app freezes and some folders down the metrics tree won't expand anymore, after loading for a while.

In the logs I can see different errors, but most notably the following ones:
`
April 4th 2018, 17:21:03.361	*** Booting runit daemon...
April 4th 2018, 17:21:03.280	*** Running /etc/rc.local...
April 4th 2018, 17:21:03.279	*** Running /etc/my_init.d/00_regen_ssh_host_keys.sh...
April 4th 2018, 17:20:58.187	*** Not all processes have exited in time. Forcing them to exit.
April 4th 2018, 17:20:53.492	04/04/2018 14:20:53 :: [listener] MetricPickleReceiver connection with 127.0.0.1:34100 lost: Connection to the other side was lost in a non-clean fashion: Connection lost.
April 4th 2018, 17:20:53.413	04/04/2018 14:20:53 :: [console] Stopping factory <twisted.internet.protocol.ServerFactory instance at 0x7ffaac913368>
April 4th 2018, 17:20:53.413	04/04/2018 14:20:53 :: [console] Stopping factory <twisted.internet.protocol.ServerFactory instance at 0x7ffaac9130e0>
April 4th 2018, 17:20:53.413	04/04/2018 14:20:53 :: [console] (TCP Port 2003 Closed)
April 4th 2018, 17:20:53.411	04/04/2018 14:20:53 :: [console] Main loop terminated.
April 4th 2018, 17:20:53.411	04/04/2018 14:20:53 :: [console] (TCP Port 2023 Closed)
April 4th 2018, 17:20:53.411	04/04/2018 14:20:53 :: [console] Server Shut Down.
April 4th 2018, 17:20:53.411	04/04/2018 14:20:53 :: [console] (TCP Port 2004 Closed)
April 4th 2018, 17:20:53.411	04/04/2018 14:20:53 :: [console] Stopping factory <twisted.internet.protocol.ServerFactory instance at 0x7ffaac913170>
April 4th 2018, 17:20:53.411	04/04/2018 14:20:53 :: [console] Stopping factory <twisted.internet.protocol.ServerFactory instance at 0x7ff61c928c68>
April 4th 2018, 17:20:53.410	04/04/2018 14:20:53 :: [console] Stopping factory <twisted.internet.protocol.ServerFactory instance at 0x7ff61c997ef0>
April 4th 2018, 17:20:53.400	04/04/2018 14:20:53 :: [console] (TCP Port 2024 Closed)
April 4th 2018, 17:20:53.398	04/04/2018 14:20:53 :: [console] Stopping factory CarbonClientFactory(127.0.0.1:2004:None)
April 4th 2018, 17:20:53.372	04/04/2018 14:20:53 :: [clients] CarbonClientFactory(127.0.0.1:2004:None)::clientConnectionLost (127.0.0.1:2004) Connection was closed cleanly.
April 4th 2018, 17:20:53.371	 - 
April 4th 2018, 17:20:53.371	    self.record_task.stop()
April 4th 2018, 17:20:53.371	04/04/2018 14:20:53 :: [clients] CarbonClientProtocol(127.0.0.1:2004:None)::connectionLost Connection was closed cleanly.
April 4th 2018, 17:20:53.371	exceptions.AssertionError: Tried to stop a LoopingCall that was not running.
April 4th 2018, 17:20:53.371	    assert self.running, ("Tried to stop a LoopingCall that was "
April 4th 2018, 17:20:53.371	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/task.py", line 171, in stop
April 4th 2018, 17:20:53.369	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 134, in maybeDeferred
April 4th 2018, 17:20:53.369	    result = f(*args, **kw)
April 4th 2018, 17:20:53.369	  File "/opt/graphite/lib/carbon/instrumentation.py", line 184, in stopService
April 4th 2018, 17:20:53.368	    l.append(defer.maybeDeferred(service.stopService))
April 4th 2018, 17:20:53.368	--- <exception caught here> ---
April 4th 2018, 17:20:53.368	  File "/usr/local/lib/python2.7/dist-packages/twisted/application/service.py", line 290, in stopService
April 4th 2018, 17:20:53.322	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 134, in maybeDeferred
April 4th 2018, 17:20:53.322	    result = f(*args, **kw)
April 4th 2018, 17:20:53.322	  File "/usr/local/lib/python2.7/dist-packages/twisted/application/service.py", line 290, in stopService
April 4th 2018, 17:20:53.322	    l.append(defer.maybeDeferred(service.stopService))
April 4th 2018, 17:20:53.321	    result = callable(*args, **kwargs)
April 4th 2018, 17:20:53.321	Traceback (most recent call last):
April 4th 2018, 17:20:53.321	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/base.py", line 407, in fireEvent
April 4th 2018, 17:20:53.319	04/04/2018 14:20:53 :: [console] (TCP Port 7002 Closed)
April 4th 2018, 17:20:53.319	04/04/2018 14:20:53 :: [console] Unhandled Error
April 4th 2018, 17:20:53.309	04/04/2018 14:20:53 :: [console] Carbon shutting down.  Changed the update rate to: 1000
April 4th 2018, 17:20:53.309	04/04/2018 14:20:53 :: [console] Received SIGTERM, shutting down.
April 4th 2018, 17:20:53.237	04/04/2018 14:20:53 :: [console] Unhandled error in Deferred:
April 4th 2018, 17:20:53.210	*** Killing all processes...
April 4th 2018, 17:20:53.210	04/04/2018 14:20:53 :: [console] Received SIGTERM, shutting down.
April 4th 2018, 17:20:53.132	*** Shutting down runit daemon (PID 8)...

`

Just before this error is thrown, I can see that nginx returned a 499 response code to my request:
`
April 4th 2018, 17:20:38.118	172.16.97.0 - - [04/Apr/2018:14:20:38 +0000] "GET /metrics/
find/?_dc=1522851518064&query=k8s.*&format=treejson&contexts=1&path=k8s&node=k8s HTTP/1.1" 499 0 "https://graphite-test.colo.elex.be/composer/?" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36"
April 4th 2018, 17:19:06.671	172.16.99.1 - guest [04/Apr/2018:14:19:06 +0000] "POST /events/ HTTP/1.1" 200 5 "-" "Apache CXF 3.0.0-milestone1"
April 4th 2018, 17:18:38.145	[Wed Apr  4 14:18:38 UTC 2018]  building index...
`
I am unsure of how to debug this further and how to improve the situation. 

Any help would be highly appreciated !

**Comment by deniszh** [comment]
Something is stopping your carbon daemon. I would say it's OOM killer but it looks like it's terminating normally, with TERM signal, like something doing `carbon stop`. 
Or maybe it's not relevant log too.

**Comment by ivanfromtheroot** [comment]
The thing is, if I empty the graphite index file, and restart the app, it works for a week or so... until the graphite index file becomes too big I guess, and times out while being queried from the web app. Does that make sense ? If yes, is there something I can do about the index growing too much ?

**Comment by deniszh** [comment]
index file is completely optional. Make it empty and root-owned - it will stop its population.

**Comment by deniszh** [comment]
Hard to say. Something is definitely wrong, but information from logs above is not enough. What's your Graphite and twisted version?

**Comment by ivaniliev359** [comment]
Twisted version is 11.1.0:

```
root@graphite-3316971257-44grr:/usr/local/src/carbon# pip show twisted
---
Name: Twisted
Version: 11.1.0
Location: /usr/local/lib/python2.7/dist-packages
Requires: zope.interface
root@graphite-3316971257-44grr:/usr/local/src/carbon# 

```
Graphite is latest one.

**Comment by deniszh** [comment]
Latest Carbon requires Twisted 13.2.0 or newer - https://github.com/graphite-project/carbon/blob/master/requirements.txt#L1 - maybe that's the case.

**Comment by ivaniliev359** [comment]
That was the cause actually... I updated Twisted python package to v17+ , along with pip 10 and Incrementals module v17+ and all started working as expected. Closing - thank you for support !

**Comment by deniszh** [comment]
👍 
Good to know, thanks!
